File: src/src/types/review.ts
// Data model types for Monocle review sessions
// See ARCHITECTURE.md section 7.2

// ---------------------------------------------------------------------------
// Union / enum types
// ---------------------------------------------------------------------------

export type ReviewState =
  | "idle"
  | "watching"
  | "reviewing"
  | "gating"
  | "submitted";

export type ReviewMode = "review" | "gate" | "hybrid";

export type CommentType = "issue" | "suggestion" | "note" | "praise";

export type CommentTargetType = "file" | "content";

export type FileStatus = "pending" | "reviewed" | "approved";

export type ChangeStatus = "added" | "modified" | "deleted" | "renamed";

export type ContentType = "markdown" | "text" | "code";

export type SubmissionAction = "approve" | "request_changes" | "comment";

// ---------------------------------------------------------------------------
// Data model
// ---------------------------------------------------------------------------

export interface DiffHunk {
  old_start: number;
  old_count: number;
  new_start: number;
  new_count: number;
  content: string;
}

export interface ReviewComment {
  id: string;
  target_type: CommentTargetType;
  target_ref: string;
  line_start: number;
  line_end: number;
  type: CommentType;
  body: string;
  code_snippet: string;
  resolved: boolean;
  created_at: number;
}

export interface ChangedFile {
  path: string;
  status: ChangeStatus;
  hunks: DiffHunk[];
  reviewed: boolean;
  comments: ReviewComment[];
}

export interface ContentItem {
  id: string;
  title: string;
  content: string;
  content_type: ContentType;
  reviewed: boolean;
  comments: ReviewComment[];
  created_at: number;
}

export interface ReviewSession {
  id: string;
  mode: ReviewMode;
  state: ReviewState;
  agent: import("./protocol").AgentName;
  repo_root: string;
  base_ref: string;
  changed_files: ChangedFile[];
  content_items: ContentItem[];
  comments: ReviewComment[];
  file_statuses: Map<string, FileStatus>;
  gate_patterns: string[];
  ignore_patterns: string[];
  created_at: number;
  updated_at: number;
}

export interface ReviewSubmission {
  id: string;
  session_id: string;
  action: SubmissionAction;
  formatted_review: string;
  comment_count: number;
  submitted_at: number;
}

File: internal/types/review.go
package types

import "time"

type AgentStatus string

const (
	AgentStatusIdle    AgentStatus = "idle"
	AgentStatusWorking AgentStatus = "working"
	AgentStatusPaused  AgentStatus = "paused"
)

type FileChangeStatus string

const (
	FileAdded    FileChangeStatus = "added"
	FileModified FileChangeStatus = "modified"
	FileDeleted  FileChangeStatus = "deleted"
	FileRenamed  FileChangeStatus = "renamed"
)

type CommentType string

const (
	CommentIssue      CommentType = "issue"
	CommentSuggestion CommentType = "suggestion"
	CommentNote       CommentType = "note"
	CommentPraise     CommentType = "praise"
)

type TargetType string

const (
	TargetFile    TargetType = "file"
	TargetContent TargetType = "content"
)

type SubmitAction string

const (
	ActionRequestChanges SubmitAction = "request_changes"
	ActionApprove        SubmitAction = "approve"
)

type ReviewSession struct {
	ID             string
	Agent          string
	AgentStatus    AgentStatus
	RepoRoot       string
	BaseRef        string
	ChangedFiles   []ChangedFile
	ContentItems   []ContentItem
	Comments       []ReviewComment
	FileStatuses   map[string]bool // path -> reviewed
	IgnorePatterns []string
	ReviewRound    int
	CreatedAt      time.Time
	UpdatedAt      time.Time
}

type ChangedFile struct {
	Path     string
	Status   FileChangeStatus
	Hunks    []DiffHunk
	Reviewed bool
	Comments []ReviewComment
}

type ContentItem struct {
	ID          string
	Title       string
	Content     string
	ContentType string
	Reviewed    bool
	Comments    []ReviewComment
	CreatedAt   time.Time
	UpdatedAt   time.Time
}

type ReviewComment struct {
	ID          string
	TargetType  TargetType
	TargetRef   string // file path or content item ID
	LineStart   int
	LineEnd     int
	Type        CommentType
	Body        string
	CodeSnippet string
	Resolved    bool
	Outdated    bool
	ReviewRound int
	CreatedAt   time.Time
	UpdatedAt   time.Time
}

type ReviewSubmission struct {
	ID              string
	SessionID       string
	Action          SubmitAction
	FormattedReview string
	CommentCount    int
	ReviewRound     int
	SubmittedAt     time.Time
}

type DiffHunk struct {
	OldStart int
	OldCount int
	NewStart int
	NewCount int
	Header   string
	Lines    []DiffLine
}

type DiffLineKind string

const (
	DiffLineContext DiffLineKind = "context"
	DiffLineAdded  DiffLineKind = "added"
	DiffLineRemoved DiffLineKind = "removed"
)

type DiffLine struct {
	Kind       DiffLineKind
	OldLineNum int
	NewLineNum int
	Content    string
}

type DiffResult struct {
	Path  string
	Hunks []DiffHunk
	Raw   string
}

type ReviewSummary struct {
	Session         *ReviewSession
	FileComments    map[string][]ReviewComment // path -> comments
	ContentComments map[string][]ReviewComment // item id -> comments
	IssueCt         int
	SuggestionCt    int
	NoteCt          int
	PraiseCt        int
	DeliveryStatus  string
}

type SubmitResult struct {
	Delivered bool
	Queued    bool
}

type SessionSummary struct {
	ID           string
	Agent        string
	RepoRoot     string
	FileCount    int
	CommentCount int
	ReviewRound  int
	CreatedAt    time.Time
	UpdatedAt    time.Time
}
